qiskit/pulse/transforms.py

```diff
@@ -317,8 +317,109 @@ def compress_pulses(schedules: List[Schedule]) -> List[Schedule]:
+            channels.
+        backend_frames: The frame configuration of the backend. Sublist i is a list of channel
+            names that belong to frame i.
+
+    Returns:
+        new_schedule: A new schedule where frames have been replaced with
+            their corresponding Drive, Control, and/or Measure channels.
+
+    Raises:
+        PulseError: if an instruction other than ShiftPhase, SetPhase, ShiftFrequency,
+            SetFrequency is applied to a Frame or if the channels in a Frame are not
+            included in the frame configuration provided by the backend.
+    """
+    if backend_frames is None:
+        return schedule
+
+    new_schedule = Schedule(name=schedule.name, metadata=schedule.metadata)
+
+    for time, inst in schedule.instructions:
+
+        # Decompose frame instructions to the channels
+        if isinstance(inst.channel, chans.Frame):
+            frame_config = set(backend_frames[inst.channel.index])
+
+            for ch in inst.channel.channels:
+                if ch.name not in frame_config:
+                    raise PulseError(f'Frame {inst.channel.name} is not compatible with '
+                                     f'the backend frames. {ch.name} not in {frame_config}.')
+
+            for ch_name in frame_config:
+                ch = _resolve_channel(ch_name)
+
+                if isinstance(inst, instructions.ShiftPhase):
+                    inst_ = instructions.ShiftPhase(inst.phase, ch)
+                elif isinstance(inst, instructions.SetPhase):
+                    inst_ = instructions.SetPhase(inst.phase, ch)
+                elif isinstance(inst, instructions.ShiftFrequency):
+                    inst_ = instructions.ShiftFrequency(inst.frequency, ch)
+                elif isinstance(inst, instructions.SetFrequency):
+                    inst_ = instructions.SetFrequency(inst.frequency, ch)
+                else:
+                    raise PulseError(inst.__class__.__name__ + ' cannot be applied to Frame.')
+
+                new_schedule.insert(time, inst_, inplace=True)
+        else:
+            new_schedule.insert(time, inst, inplace=True)
+
+    return new_schedule
+
+
+def _resolve_channel(name: str) -> chans.Channel:
+    """
+    Helper method to resolve the channels in the frame.
+
+    Args:
+        name: Name of the channel.
+
+    Returns:
+        channel: The channel of the appropriate type.
+
+    Raises:
+        PulseError: If the channel is not a DriveChannel, ControlChannel, or
+            MeasureChannel.
+    """
+    if name[0] == 'd':
+        return chans.DriveChannel(int(name[1:]))
+    elif name[0] == 'u':
+        return chans.ControlChannel(int(name[1:]))
+    elif name[0] == 'm':
+        return chans.MeasureChannel(int(name[1:]))
+
+    raise PulseError('Channel %s not supported in frames.' % name)
+
+
+def _flatten_frames(channels: Tuple[chans.Channel]) -> Set[chans.Channel]:
```

Comment:
Can we internally call this in `Schedule`? i.e. `Schedule._mutable_insert` is good place to call this function. This way we can always have unified channel representation in schedule. Currently `timeslots` for frames will be created but I think this is not sensible.